File: graph/KruskalMinimumSpanningTree/cpp/KruskalMST.cpp
#include "KruskalMST.hpp"

#include <algorithm>
#include <limits>
#include <numeric>

namespace {

struct DisjointSetUnion {
    std::vector<int> parent;
    std::vector<int> size;

    explicit DisjointSetUnion(int nodeCount) : parent(nodeCount), size(nodeCount, 1) {
        std::iota(parent.begin(), parent.end(), 0);
    }

    int Find(int node) {
        if (parent[node] == node) {
            return node;
        }
        parent[node] = Find(parent[node]);
        return parent[node];
    }

    bool Unite(int left, int right) {
        left = Find(left);
        right = Find(right);
        if (left == right) {
            return false;
        }
        if (size[left] < size[right]) {
            std::swap(left, right);
        }
        parent[right] = left;
        size[left] += size[right];
        return true;
    }
};

bool CheckedAdd(long long left, long long right, long long& out) {
    if (right > 0 && left > std::numeric_limits<long long>::max() - right) {
        return false;
    }
    if (right < 0 && left < std::numeric_limits<long long>::min() - right) {
        return false;
    }
    out = left + right;
    return true;
}

Edge Normalize(const Edge& edge) {
    if (edge.from <= edge.to) {
        return edge;
    }
    return {edge.to, edge.from, edge.weight};
}

bool EdgeLess(const Edge& left, const Edge& right) {
    if (left.weight != right.weight) {
        return left.weight < right.weight;
    }
    if (left.from != right.from) {
        return left.from < right.from;
    }
    return left.to < right.to;
}

}  // namespace

KruskalResult KruskalMST(int nodeCount, const std::vector<Edge>& edges) {
    std::vector<Edge> sortedEdges;
    sortedEdges.reserve(edges.size());
    for (const Edge& edge : edges) {
        sortedEdges.push_back(Normalize(edge));
    }
    std::sort(sortedEdges.begin(), sortedEdges.end(), EdgeLess);

    DisjointSetUnion dsu(nodeCount);
    KruskalResult result{KruskalStatus::Ok, 0, {}};
    result.chosenEdges.reserve(nodeCount > 0 ? nodeCount - 1 : 0);

    for (const Edge& edge : sortedEdges) {
        if (!dsu.Unite(edge.from, edge.to)) {
            continue;
        }

        long long updatedWeight = 0;
        if (!CheckedAdd(result.totalWeight, edge.weight, updatedWeight)) {
            result.status = KruskalStatus::Overflow;
            return result;
        }

        result.totalWeight = updatedWeight;
        result.chosenEdges.push_back(edge);

        if (static_cast<int>(result.chosenEdges.size()) == nodeCount - 1) {
            return result;
        }
    }

    result.status = KruskalStatus::Disconnected;
    return result;
}

File: graph/KruskalMinimumSpanningTree/java/KruskalMST.java
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public final class KruskalMST {
    private KruskalMST() {}

    public static final class Edge {
        public final int source;
        public final int target;
        public final long weight;

        public Edge(int source, int target, long weight) {
            this.source = source;
            this.target = target;
            this.weight = weight;
        }
    }

    public enum Status {
        OK,
        DISCONNECTED,
        OVERFLOW
    }

    public static final class Result {
        public final Status status;
        public final long totalWeight;
        public final List<Edge> chosenEdges;

        public Result(Status status, long totalWeight, List<Edge> chosenEdges) {
            this.status = status;
            this.totalWeight = totalWeight;
            this.chosenEdges = chosenEdges;
        }
    }

    private static final class DisjointSetUnion {
        private final int[] parent;
        private final int[] size;

        private DisjointSetUnion(int nodeCount) {
            this.parent = new int[nodeCount];
            this.size = new int[nodeCount];
            for (int node = 0; node < nodeCount; ++node) {
                this.parent[node] = node;
                this.size[node] = 1;
            }
        }

        private int find(int node) {
            if (parent[node] == node) {
                return node;
            }
            parent[node] = find(parent[node]);
            return parent[node];
        }

        private boolean unite(int left, int right) {
            left = find(left);
            right = find(right);
            if (left == right) {
                return false;
            }
            if (size[left] < size[right]) {
                int temporary = left;
                left = right;
                right = temporary;
            }
            parent[right] = left;
            size[left] += size[right];
            return true;
        }
    }

    private static Long checkedAdd(long left, long right) {
        if (right > 0 && left > Long.MAX_VALUE - right) {
            return null;
        }
        if (right < 0 && left < Long.MIN_VALUE - right) {
            return null;
        }
        return left + right;
    }

    private static Edge normalize(Edge edge) {
        if (edge.source <= edge.target) {
            return edge;
        }
        return new Edge(edge.target, edge.source, edge.weight);
    }

    public static Result kruskalMST(int nodeCount, List<Edge> edges) {
        List<Edge> sortedEdges = new ArrayList<>(edges.size());
        for (Edge edge : edges) {
            sortedEdges.add(normalize(edge));
        }
        sortedEdges.sort(
                Comparator.comparingLong((Edge edge) -> edge.weight)
                        .thenComparingInt(edge -> edge.source)
                        .thenComparingInt(edge -> edge.target));

        DisjointSetUnion dsu = new DisjointSetUnion(nodeCount);
        List<Edge> chosenEdges = new ArrayList<>(Math.max(0, nodeCount - 1));
        long totalWeight = 0;

        for (Edge edge : sortedEdges) {
            if (!dsu.unite(edge.source, edge.target)) {
                continue;
            }

            Long updated = checkedAdd(totalWeight, edge.weight);
            if (updated == null) {
                return new Result(Status.OVERFLOW, totalWeight, chosenEdges);
            }

            totalWeight = updated;
            chosenEdges.add(edge);

            if (chosenEdges.size() == nodeCount - 1) {
                return new Result(Status.OK, totalWeight, chosenEdges);
            }
        }

        return new Result(Status.DISCONNECTED, totalWeight, chosenEdges);
    }
}
